### PR TITLE
Patch callchain samples.

### PR DESCRIPTION
**Pull request by florian-kuebler**
Patches callchain samples that fall into uprobes and always remove the top of the callstack, as this falls into the kernel.

### CODE REVIEW

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp [hunk 1]
```diff
@@ -145,7 +147,9 @@ UprobesUnwindingVisitor::CallstackFramesFromInstructionPointers(
     return std::vector<CallstackFrame>();
   }
   std::vector<CallstackFrame> callstack_frames;
-  for (uint64_t i = 0; i < size; i++) {
+  // The top of a callchain is always inside the kernel code.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp:150]: Is this the right function for this? This function performs the transformation to a `std::vector<CallstackFrame>`. It seems that other kinds of processing should be done by the caller or somewhere else.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 2]
```diff
@@ -59,6 +59,34 @@ class UprobesReturnAddressManager {
     }
   }
 
+  void PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t nr,
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:64]: You should add tests. Indeed, this class has tests.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 3]
```diff
@@ -59,6 +59,34 @@ class UprobesReturnAddressManager {
...
+    if (!tid_uprobes_stacks_.contains(tid)) {
+      return;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    uint64_t i = 0;
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
+    //  hijacked an address at the same stack.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:75]:
```suggestion
    //  hijacked an address at the same stack pointer?
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 4]
```diff
@@ -59,6 +59,34 @@ class UprobesReturnAddressManager {
...
+                      unwindstack::Maps* maps) {
+    if (!tid_uprobes_stacks_.contains(tid)) {
+      return;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    uint64_t i = 0;
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:74]: Have you tried what happens? Maybe in such case the second uretprobe (the first to be encountered in the loop) shouldn't be applied? In other words, maybe when there are consecutive `OpenUprobes`s with the same stack pointer, only the outer one should be applied (just guessing here)? A basic solution might end up being quite easy.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 5]
```diff
@@ -59,6 +59,34 @@ class UprobesReturnAddressManager {
...
+    //  hijacked an address at the same stack.
+    for (auto it = tid_uprobes_stack.rbegin(); it != tid_uprobes_stack.rend();
+         it++) {
+      const OpenUprobes& uprobes = *it;
+      for (; i < nr; i++) {
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:79]: Could you write this with a single loop where you go through the callchain and increment `it` when you encounter a uprobes IP? In other words, iterate over the callchain, not over `tid_uprobes_stack`? I think it would be easier to understand.

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp (hunk 1 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesUnwindingVisitor.cpp:150]: Maybe just 
```
return_address_manager_.PatchCallchain(event->GetTid(), event->GetCallchain() + 1,
                                         event->GetCallchainSize() - 1,
                                         current_maps_.get());
```
with a comment explaining it (assuming it's already checked somewhere that the callchain is not empty).

Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Thanks. I moved the stripping of the callchain to the caller, as you suggested.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h (hunk 5 above)
Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h]: I refactored it like that. Hope it is better now.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 6]
```diff
@@ -59,6 +61,38 @@ class UprobesReturnAddressManager {
     }
   }
 
+  void PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t nr,
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:64]: Maybe rename `nr` to something more descriptive like `callchain_size`.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 7]
```diff
@@ -59,6 +61,38 @@ class UprobesReturnAddressManager {
...
+    CHECK(!tid_uprobes_stack.empty());
+
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
+    //  hijacked an address at the same stack pointer.
+    auto uprobes_it = tid_uprobes_stack.rbegin();
+    for (uint64_t i = 0; i < nr; i++) {
+      uint64_t ip = callchain[i];
+      unwindstack::MapInfo* map_info = maps->Find(ip);
+
+      // Only patch Broken IPs
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:80]: Maybe remove this comment and add a comment for the function describing briefly what is being done (as all the comments related to this class are for "stack patching" but not for "callchain patching").

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 8]
```diff
@@ -59,6 +61,38 @@ class UprobesReturnAddressManager {
...
+      if (map_info == nullptr || map_info->name != "[uprobes]") {
+        continue;
+      }
+
+      const OpenUprobes& uprobes = *uprobes_it;
+      callchain[i] = uprobes.return_address;
+      uprobes_it++;
+
+      // There can't be more frames to patch as uprobes.
+      if (uprobes_it == tid_uprobes_stack.rend()) {
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:90]: What about checking this before the actual patching and printing an error if this is true?

Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h:90]: I was thinking about this first. The point here is that, if we have a callchain of lets say 100 frames, and there is just one uprobe, which is very much at the top of the chain, we can exit the loop quite early, as there can't be any other uprobe.

Comment (dpallotti) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h:90]: Your point makes sense. Personally I think that for now it could be useful to have such errors while you make experiments, you can add a TODO explaining how that can be optimized later.
I would even add the opposite error, that is if after the loop `uprobes_it == tid_uprobes_stack.rend()` doesn't hold.
All this is especially since this code doesn't really have other ways to check that the correct frames are being patches.
In general and in other words, I would have some kind of error print if the number of uprobes frames in the callchain is different than the size of `tid_uprobes_stack`.

**Comment by florian-kuebler** [comment]
I added handling for a couple of corner cases and added tests.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 9]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
     }
   }
 
+  // In case of callchain sampling we don't have the complete stack to patch,
+  // but only the callchain (as list of instruction pointers). In those,
+  // a uprobe address occurs instead of the caller of an instrumented function.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:66]:
```suggestion
  // a uprobe address occurs in place of the caller of an instrumented function.
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 10]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+  // This function patches the callchain, using the maps information to identify
+  // instruction pointers of uprobe code and using the return address safed in
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:68]:
```suggestion
  // instruction pointers of uprobe code and using the return address saved in
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 11]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+      }
+
+      ips_to_patch.push_back(i);
+    }
+
+    // In case we already used all uprobes, we need to discard this sample.
+    // There are two situations where this may happen:
+    //  1. At the beginning of a capture, where we missed the first uprobes
+    //  2. When there are two many events, resulting in lost events or
+    //   processing out of order.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:98]:
```suggestion
    //  2. When some events are lost or processed out of order.
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 12]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+    if (tid_uprobes_stack.size() < ips_to_patch.size()) {
+      ERROR(
+          "Discarding sample in an uprobe as some uprobe records are missing.");
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:101]:
```suggestion
          "Discarding sample in a uprobe as some uprobe records are missing.");
```
This kind of "u" at the beginning of a word is a consonant sound.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 13]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+      return false;
+    }
+    // In cases of lost events, or out of order processing, there might be wrong
+    //  uprobes. So we need to discard the event. In general we should be fast
+    //  enough, such that this does not happen.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:106]:
```suggestion
    // uprobes. So we need to discard the event. In general we should be fast
    // enough, such that this does not happen.
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 14]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+      return false;
+    }
+
+    auto uprobes_it = tid_uprobes_stack.rbegin();
+    // There are two situations where this may happen:
+    //  1. At the very end of an instrumented function, where the return
+    //   address was already restored.
+    //  2. At the very beginning of an instrumented function, where the return
+    //   address was not yet overridden.
+    // In any case, the inner most uprobe has not overridden the return address.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:118]:
```suggestion
    // In any case, the innermost uprobe has not overridden the return address.
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 15]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+    // We do not need to patch the affect of this uprobe and can move forward.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:119]:
```suggestion
    // We do not need to patch the effect of this uprobe and can move forward.
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 16]
```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
...
+  bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
+                      unwindstack::Maps* maps) {
+    if (!tid_uprobes_stacks_.contains(tid)) {
+      return true;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    std::vector<uint64_t> ips_to_patch;
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:79]: The content is not IPs, I would call it something more generic like `frames_to_patch`.

File: OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp [hunk 17]
```diff
@@ -390,4 +392,228 @@ TEST(UprobesReturnAddressManager, TailCallOptimization) {
   test_handler.OnNonUretprobesReturn();
 }
 
+//==============================================================================
+// Tests for frame pointer based callchains start here:
+//==============================================================================
+
+const std::string maps_string =
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp:399]: This should be static or inside an anonymous namespace.

File: OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp [hunk 18]
```diff
@@ -390,4 +392,228 @@ TEST(UprobesReturnAddressManager, TailCallOptimization) {
...
+    "7ffcae624000-7ffcae646000 rw-p 00000000 00:00 0                          "
+    "[stack]\n"
+    "7ffcae7f0000-7ffcae7f3000 r--p 00000000 00:00 0                          "
+    "[vvar]\n"
+    "7ffcae7f3000-7ffcae7f4000 r-xp 00000000 00:00 0                          "
+    "[vdso]\n"
+    "7fffffffe000-7ffffffff000 --xp 00000000 00:00 0                          "
+    "[uprobes]";
+
+auto maps = LibunwindstackUnwinder::ParseMaps(maps_string);
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp:481]: Static or anonymous namespace (also please specify the type).

File: OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp [hunk 19]
```diff
@@ -390,4 +392,228 @@ TEST(UprobesReturnAddressManager, TailCallOptimization) {
...
+
+TEST(UprobesReturnAddressManager, CallchainNoUprobes) {
+  UprobesReturnAddressManager return_address_manager;
+
+  std::vector<uint64_t> expected_callchain{
+      18446744073709551104lu, 94355907990078lu, 94355907990120lu,
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp:487]: Soooo, since these values are "real" instruction pointer values, why didn't you use the hexadecimal notation? That way it would be possible to make a parallel with the maps, while like this they look like random numbers.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h (hunk 16 above)
Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h]: Sure

File: OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp (hunk 17 above)
Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp:400]: I added an anonymous namespace.

File: OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp (hunk 19 above)
Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp]: Converted all values to hex.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 20]
```diff
@@ -59,6 +61,74 @@ class UprobesReturnAddressManager {
...
+    if (tid_uprobes_stack.size() < frames_to_patch.size()) {
+      ERROR(
+          "Discarding sample in a uprobe as some uprobe records are missing.");
+      return false;
+    }
+    // In cases of lost events, or out of order processing, there might be wrong
+    // uprobes. So we need to discard the event. In general we should be fast
+    // enough, such that this does not happen.
+    if (tid_uprobes_stack.size() > frames_to_patch.size() + 1) {
+      ERROR("Discarding sample in an uprobe as uprobe records are incorrect.");
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:107]:
```suggestion
      ERROR("Discarding sample in a uprobe as uprobe records are incorrect.");
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h (hunk 4 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h:81]: Fine, but don't consider this a rare corner case, tail call optimization is common.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 21]
```diff
@@ -59,6 +61,74 @@ class UprobesReturnAddressManager {
...
+      return true;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    std::vector<uint64_t> frames_to_patch;
+
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
+    //  hijacked an address at the same stack pointer.
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:82]:
```suggestion
    //  hijacked an address at the same stack pointer?
```

File: OrbitLinuxTracing/UprobesReturnAddressManager.h [hunk 22]
```diff
@@ -59,6 +61,74 @@ class UprobesReturnAddressManager {
...
+    // In case we already used all uprobes, we need to discard this sample.
+    // There are two situations where this may happen:
+    //  1. At the beginning of a capture, where we missed the first uprobes
+    //  2. When some events are lost or processed out of order.
+    if (tid_uprobes_stack.size() < frames_to_patch.size()) {
+      ERROR(
+          "Discarding sample in a uprobe as some uprobe records are missing.");
+      return false;
+    }
+    // In cases of lost events, or out of order processing, there might be wrong
```
Comment (dpallotti) [inline OrbitLinuxTracing/UprobesReturnAddressManager.h:103]: Are lost events or out of order processing the causes of `tid_uprobes_stack.size() > frames_to_patch.size() + 1`? This seems more like a case that shouldn't happen and that if/when does happen the reason could be unclear.

Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h:103]: Well, what happen for me was, that there was new uprobes before the uretprobes of the previous uprobe.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h (hunk 4 above)
Comment (florian-kuebler) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h:81]: Right, I'll add a bug for this and will come back as to this as soon as possible. For now, need to re-prioritize.

File: OrbitLinuxTracing/UprobesReturnAddressManager.h (hunk 22 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesReturnAddressManager.h:103]: Ah, makes sense, so basically when a uretprobe is lost or late. (This is an unfortunate case that hopefully doesn't happen on release builds because if it caused exactly `tid_uprobes_stack.size() > frames_to_patch.size() + 1` then it would fall in the other case and would be handled incorrectly.)

**Comment by florian-kuebler** [comment]
@dpallotti Anything from your side, before merging this?

**Comment by dpallotti** [comment]
> @dpallotti Anything from your side, before merging this?

Looks good, go ahead!
